src/core/main/common/FreezeResolver.ts

```diff
@@ -0,0 +1,573 @@
+            }
+          }
+        }
+      }
+    }
+
+    if (shouldHandleDecipherabilityFreeze && !isClear && hasOnlyDecipherableSegments) {
+      log.warn(
+        "FR: we are frozen despite only having decipherable " +
+          "segments left in the buffer, reloading",
+      );
+      this._decipherabilityFreezeStartingTimestamp = null;
+      this._ignoreFreezeUntil = now + MINIMUM_TIME_BETWEEN_FREEZE_HANDLING;
+      return { type: "reload", value: null };
+    }
+    return null;
+  }
+
+  /**
+   * This method should only be called if a "flush" strategy has recently be
+   * taken to try to unfreeze playback yet playback is still frozen.
+   *
+   * It considers the current played content and returns a more-involved
+   * unfreezing strategy (most often reload-related) to try to unfree playback.
+   * @param {number} freezingPosition - The playback position at which we're
+   * currently frozen.
+   * @returns {Object}
+   */
+  private _getStrategyIfFlushingFails(freezingPosition: number): IFreezeResolution {
+    log.warn(
+      "FR: A recent flush seemed to have no effect on freeze, checking for transitions",
+    );
+
+    /** Contains Representation we might want to avoid after the following algorithm */
+    const toAvoid = [];
+
+    for (const ttype of ["audio", "video"] as const) {
+      const segmentList = this._lastSegmentInfo[ttype];
+      if (segmentList.length === 0) {
+        // There's no buffered segment for that type, go to next type
+        continue;
+      }
+
+      /** Played history information on the current segment we're stuck on. */
+      let currentSegmentEntry = segmentList[segmentList.length - 1];
+      if (currentSegmentEntry.segment === null) {
+        // No segment currently played for that given type, go to next type
+        continue;
+      }
+
+      /** Metadata on the segment currently being played. */
+      const currentSegment = currentSegmentEntry.segment;
+
+      /**
+       * Set to the first previous segment which is linked to a different
+       * Representation.
+       */
+      let previousRepresentationEntry: IPlayedHistoryEntry | undefined;
+
+      // Now find `previousRepresentationEntry` and `currentSegmentEntry`.
+      for (let i = segmentList.length - 2; i >= 0; i--) {
+        const segment = segmentList[i];
+        if (segment.segment === null) {
+          // Before the current segment, there was no segment being played
+          previousRepresentationEntry = segment;
+          break;
+        } else if (
+          segment.segment.infos.representation.uniqueId !==
+            currentSegment.infos.representation.uniqueId &&
+          currentSegmentEntry.timestamp - segment.timestamp < 5000
+        ) {
+          // Before the current segment, there was a segment of a different
+          // Representation being played
+          previousRepresentationEntry = segment;
+          break;
+        } else if (
+          currentSegment !== null &&
```

Comment:
is it a necessary check? from what I read currentSegment can never be `null` because we check for it line 374.

Reply:
I think you're right, shamefully not the easiest code to follow :D
Done